### PR TITLE
[RFC] docs/design: proposal to determine application protocol

### PR DESCRIPTION
**Pull request by shashankram**
**Description**:
This change adds a proposal doc to address the need to be able
to determine the application protocol served by a service's port.

Part of #1521

Signed-off-by: Shashank Ram <shashank08@gmail.com>

**Affected area**:

- New Functionality      [ ]
- Documentation          [X]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
`No`

### CODE REVIEW

**Review by eduser25 — DISMISSED** [review]

File: docs/design/service_app_protocol.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Application Protocol Selection
+
+Kubernetes services expose one or more ports. A port exposed by an application running the service can serve a specific application protocol such as HTTP, TCP etc. There is a requirement to be able to determine the application protocol served by a service’s port to be able to correctly configure protocol specific configurations in Envoy (Filters, Routes, RBAC policies). Prior to Kubernetes 1.19, there is no way to determine the application protocol for a port. The IP `protocol` field on the Service resource must be one of TCP, UDP, or SCTP, so this cannot be used to determine the application protocols such HTTP, gRPC etc. To address this limitation, Kubernetes 1.19 introduced an [`AppProtocol`](https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol) field for the Service resource.
```
Comment (draychev) [inline docs/design/service_app_protocol.md:3]: @shashankram would you be so kind to give an example.  You and I discussed how at first it appears that SMI spec may have enough context for OSM Controller do determine the protocol.  But looking deeper it appears that we need extra context.

I'd give an example of SMI Spec + Service + AppProtocol YAML.

File: docs/design/service_app_protocol.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    name: some-port
+    appProtocol: http
+```
+
+A service `service-2` backed by a pod in service account `sa-1` serving raw `tcp` application traffic shold be defined as follows:
+
+```yaml
+kind: Service
+metadata:
+  name: service-1
```
Comment (draychev) [inline docs/design/service_app_protocol.md:77]: This is `service-2` right?

```suggestion
  name: service-2
```

File: docs/design/service_app_protocol.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  namespace: default
+spec:
+  ports:
+  - port: 8080
+    name: some-port
+    appProtocol: tcp
```
Comment (draychev) [inline docs/design/service_app_protocol.md:83]:
```suggestion
    appProtocol: tcp # Same port number as service-1; different protocol
```

File: docs/design/service_app_protocol.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+```yaml
+kind: Service
+metadata:
+  name: service-1
+  namespace: default
+spec:
+  ports:
+  - port: 8080
+    name: some-port
+    appProtocol: http
```
Comment (draychev) [inline docs/design/service_app_protocol.md:69]:
```suggestion
    appProtocol: http # Same port number as service-2; different protocol
```

File: docs/design/service_app_protocol.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    - GET
+---
+kind: TrafficTarget
+metadata:
+  name: test
+  namespace: default
+spec:
+  destination:
+    kind: ServiceAccount
+    name: sa-1
```
Comment (draychev) [inline docs/design/service_app_protocol.md:43]:
```suggestion
    name: sa-1 # There are 2 services under this service account:  service-1 and service-2
```

**Review by draychev — APPROVED** [review]
@shashankram  thanks for adding the examples!
I added a few comments to the YAML.
LGTM

File: docs/design/service_app_protocol.md (hunk 3 above)
Comment (shashankram) [reply docs/design/service_app_protocol.md:83]: This need not be true. I don't want to be introducing such assumptions as a part of this PR.

**Review by draychev — APPROVED** [review]
👍

**Review by eduser25 — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]

**Review by michelleN — APPROVED** [review]

**Pull request merged**